Add the possibility to use a system wide installation of tl-optional

GiulioRomualdi opened the pull request:
Hi @artivis, this PR adds the possibility to use a system-wide installation of tl-optional. In details:
- rename `external/lt/lt/optional.hpp → external/tl/tl/optional.hpp` (this is required to be compliant with the official installation of tl-optional)
- `tl-optional` is now searched with `find_package(tl-optional QUIET)`. If not found the version provided by `manif` is used and installed.
- I also bumped `tl-optional` to the latest version (1.0.0).

I tested the PR on my setup (ubuntu 20.04) with `tl-optional` installed. The tests were passing and the python bindings correctly generated.

This PR will fix the problem of porting manif in vcpkg spotted by @traversaro in https://github.com/microsoft/vcpkg/issues/16058

GiulioRomualdi commented:
https://github.com/artivis/manif/pull/207 should allow us to use tl-optional provided by vcpkg.

traversaro commented:
> tl-optional is now searched with find_package(tl-optional QUIET). If not found the version provided by manif is used and installed.

Can you provide a way (option or what do you prefer, even disabled by default) to have a way in which if tl-optional is not found the build simply fails? Because for packaging this is the kind of behavior that is convenient to use. See https://www.youtube.com/watch?v=sBP17HQAQjk&t=654 for more details

artivis commented:
@GiulioRomualdi the failing Ubuntu 20.04 CI job *may be* related to [this fix not being ported](https://github.com/artivis/manif/commit/2133b390ecfff794787f3057fdfa8c4df437a72e#diff-7b39518b1c09b7999da1c692a959ca461c9189ef75d3d60049373e8941df1cb3) after updating `optional` to 1.0.0.

artivis commented on CMakeLists.txt:
```diff
@@ -77,10 +77,27 @@ add_library(${PROJECT_NAME} INTERFACE)
+# Check if there exists a system wise installation of tl-optional
+find_package(tl-optional QUIET)
+
+# if tl-option exists manif will use the system wise installation otherwise it will provide it's own version
+if(tl-optional_FOUND)
+
+  target_include_directories(${PROJECT_NAME} INTERFACE
+    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include>"
+    )
+
+  target_link_libraries(${PROJECT_NAME} INTERFACE tl::optional)
+
+else()
+
+  target_include_directories(${PROJECT_NAME} INTERFACE
+    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include;${PROJECT_SOURCE_DIR}/external/tl>"
+    "$<INSTALL_INTERFACE:include>"
+    )
+
+endif()
```
The following is a little more concise. Not tested with an actual system-installed tl-optional,

```suggestion
target_include_directories(${PROJECT_NAME} INTERFACE
  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include;$<$<NOT:$<BOOL:tl-optional_FOUND>>:${PROJECT_SOURCE_DIR}/external/tl>>"
  "$<INSTALL_INTERFACE:include>"
)

# if tl-option exists manif will use the system wise installation otherwise it will provide it's own version
if(tl-optional_FOUND)
  target_link_libraries(${PROJECT_NAME} INTERFACE tl::optional)
endif()
```

artivis commented on CMakeLists.txt:
```diff
@@ -151,10 +168,14 @@ install(
+  install(
+    FILES "${PROJECT_SOURCE_DIR}/external/tl/tl/optional.hpp"
+    DESTINATION include/tl
+    )
```
```suggestion
  )
```

artivis commented:
Any idea how things goes if one tries to install `tl-optional` after `manif` installed the header?

GiulioRomualdi commented:
> @GiulioRomualdi the failing Ubuntu 20.04 CI job _may be_ related to [this fix not being ported](https://github.com/artivis/manif/commit/2133b390ecfff794787f3057fdfa8c4df437a72e#diff-7b39518b1c09b7999da1c692a959ca461c9189ef75d3d60049373e8941df1cb3) after updating `optional` to 1.0.0.

Hi @artivis as far as I understood the original tl-optional does not require that fixes so probably something strange is happening in the example. (Im investigating on this)

artivis commented:
Hi,
I'm not sure what you mean by 

> as far as I understood the original tl-optional does not require that fixes

But the error in Ubuntu 20.04 CI is 

```bash
error: base class ‘struct tl::detail::optional_operations_base<Eigen::Ref<Eigen::Matrix<double, 9, 9>, 0, Eigen::OuterStride<> > >’ should be explicitly initialized in the copy constructor [-Werror=extra]
```

which is what the aforementioned patch fixes. Note that it is activated by the  CMake flag `-Werror=extra`.
It should probably be fixed upstream too.

artivis commented:
@GiulioRomualdi hey, any chance to see this going through anytime soon?

GiulioRomualdi commented:
Hi @artivis https://github.com/TartanLlama/optional/pull/45 should fix the problem

traversaro commented:
> 
> 
> Hi @artivis [TartanLlama/optional#45](https://github.com/TartanLlama/optional/pull/45) should fix the problem

For anyone that was subscribed to this PR and not to that one, that PR was merged two days ago.

GiulioRomualdi commented:
Hi @artivis and @traversaro. 

I  bumped tl-optional stored in the `external` folder to the latest version of master  TartanLlama/optional@c28fcf7. I also changed the CMake machinery applying what @traversaro suggested in https://github.com/artivis/manif/pull/207#issuecomment-790784893 and what @artivis suggested in https://github.com/artivis/manif/pull/207#discussion_r587746978


In details, I introduced `USE_SYSTEM_WISE_TL_OPTIONAL` parameter (default value `OFF`). If the parameter is `ON`, `cmake` search for `tl-optional`.

artivis commented on CMakeLists.txt:
```diff
@@ -77,10 +77,17 @@ add_library(${PROJECT_NAME} INTERFACE)
 #  INTERFACE_COMPILE_FEATURES cxx_std_11
 #)
 
+option(USE_SYSTEM_WISE_TL_OPTIONAL "Use a system wise installation of tl-optional" OFF)
```
Is it 'system wide' or 'system wise' ? Not sure myself.

GiulioRomualdi replied:
I think It should be systemwide. 
https://en.wiktionary.org/wiki/systemwide

I  fix the typo

artivis commented on CMakeLists.txt:
```diff
@@ -77,10 +77,17 @@ add_library(${PROJECT_NAME} INTERFACE)
 
+option(USE_SYSTEM_WISE_TL_OPTIONAL "Use a system wise installation of tl-optional" OFF)
+
+if(USE_SYSTEM_WISE_TL_OPTIONAL)
+  find_package(tl-optional REQUIRED)
+endif()
```
A system installation of `tl-optional` is *only* used when this option is set. I'd rather prefer that it is used whenever available. Something along the lines of,
```suggestion
find_package(tl-optional 1.0.0 QUIET)
if(USE_SYSTEM_WISE_TL_OPTIONAL AND NOT tl-optional_FOUND)
  message(FATAL_ERROR "Could not find required package tl-optional")
endif()
```

artivis commented on CMakeLists.txt:
```diff
@@ -77,10 +77,17 @@ add_library(${PROJECT_NAME} INTERFACE)
+
 target_include_directories(${PROJECT_NAME} INTERFACE
-  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include;${PROJECT_SOURCE_DIR}/external/lt>"
+  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include$<$<NOT:$<BOOL:${USE_SYSTEM_WISE_TL_OPTIONAL}>>:;${PROJECT_SOURCE_DIR}/external/tl>>"
```
As per my previous comment, use `tl-optional_FOUND` instead of `USE_SYSTEM_WISE_TL_OPTIONAL`

GiulioRomualdi replied:
If `tl-optional_FOUND` is true and https://github.com/artivis/manif/pull/207#discussion_r649952287 is applied a systemwide installation of tl-optional will be used even if `USE_SYSTEM_WIDE_TL_OPTIONAL` is off. 

Is this one the desired behavior?

artivis replied:
If the system tl-optional is at least v.1, then I'd say yes this is the desirable behavior.

artivis commented on CMakeLists.txt:
```diff
@@ -151,10 +158,14 @@ install(
-)
+
+# if tl-optional is not found manif will install its own version
+if(NOT USE_SYSTEM_WISE_TL_OPTIONAL)
```
`USE_SYSTEM_WISE_TL_OPTIONAL`  -> `tl-optional_FOUND`

artivis commented on CMakeLists.txt:
```diff
@@ -77,10 +77,17 @@ add_library(${PROJECT_NAME} INTERFACE)
+  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include$<$<NOT:$<BOOL:${USE_SYSTEM_WISE_TL_OPTIONAL}>>:;${PROJECT_SOURCE_DIR}/external/tl>>"
   "$<INSTALL_INTERFACE:include>"
 )
+target_link_libraries(${PROJECT_NAME} INTERFACE "$<$<BOOL:${USE_SYSTEM_WISE_TL_OPTIONAL}>:tl::optional>")
```
`USE_SYSTEM_WISE_TL_OPTIONAL`  -> `tl-optional_FOUND`

GiulioRomualdi commented:
Hi @artivis  in https://github.com/artivis/manif/pull/207/commits/27e47a16c5a28e4eb96c446b147df51cf97e70a4 I applied your suggestions. Let me know if I should do something more

artivis reviewed (approved):
LGTM :+1:. Thanks @GiulioRomualdi
